WordPress/gutenberg
# Allow block use based on post-type or template context

⚠️ This PR is a POC



## What?


This PR allows making the block visible only on specific post types and/or specific templates. Fix #41062.


## Why?



The current way to hide blocks based on context is a little bit tricky (how to do this is in the description of #41062). For `WooCommerce Blocks`, we need this feature because we want to enable specific blocks only for a specific template (https://github.com/woocommerce/woocommerce-blocks/issues/5193). We tried as suggested (https://github.com/woocommerce/woocommerce-blocks/pull/5637) on #41062, but we would prefer to have a better way to do this on the Gutenberg side.

## How?


I'm pretty new to Gutenberg codebase: I noticed that all the logic around the blocks and the `Inserter Block` is contained in `@wordpress/block-editor` package. The filter that I'm trying to add has to be aware of WordPress context and in accordance with the [modularity documentation](https://github.com/WordPress/gutenberg/blob/f8dc76bff9f817d12771c6ae75f26aaa0d57844a/docs/explanations/architecture/modularity.md/#L89) this package is independent of the WordPress ecosystem.

This is the only way that I found to implement this feature without a huge refactor. I'm not sure that is the correct way to go, so I'm open to feedback and hints :) 

With this implementation, there are some use cases that I want to highlight:
- A saved post/page/template uses a block that, after an update to the value `allowForPostTypes` or `allowForTemplates` it should be not possible to add on that post/page/template, the added block will be still visible, but if the user removes the block, this latter will not be able to re-added.
- A pattern uses a block that it is possible to add only on a specific post type: what should happen when the user adds the pattern on a different post type? Should the block be added?




## Testing Instructions


In this PR, there are changes on the `block.json` of the `Code Block`. This is just to speed up the testing, but obviously, these changes will not be included in the final code. In accordance with the changes the block it should be visible only on:
- page
- archive template

On the WordPress codebase, I added in [this array](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks.php#L284), these two lines:

```
"allowForPostTypes" => 'allow_for_post_types',
"allowForTemplates" => 'allow_for_templates',
```

After that check out this branch:

1. Create a new post.
2. Search for `Code` block: it should be not visible.
3. Create a new page.
4. Search for `Code` block: it should be visible.
5. Open the FSE.
6. Open the `Home` template.
7.  Search for `Code` block: it should be not visible.
8. Open the `Archive` template .
9. Search for `Code` block: it should be visible.
